### PR TITLE
Fix memory leak

### PR DESCRIPTION
**Pull request by Khang-NT**

### CODE REVIEW

File: app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java [hunk 1]
```diff
@@ -216,8 +216,13 @@ public static void clearMostPlayed() {
         void onSave(Playlist playlist);
     }
 
-    public static void makePlaylistMenu(SubMenu sub) {
-        DataManager.getInstance().getPlaylistsRelay()
+    public static void makePlaylistMenu(SubMenu sub, Observable<Long> stopEventTrigger, boolean autoUpdate) {
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java]: It looks like this was just an oversight - the Observable remains subscribed and listening for emissions after we've finished making our playlist menu. We only need the first emission from the playlist relay.

Using the stopEventTrigger/LifecycleCallbacks seems like overkill here. I would think instead, we should just be calling `playlistListObservable.first(Collections.emptyList())`, which ensures the `Observable` only emits once and then is disposed.

File: app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerView.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 public interface DrawerView extends PurchaseView {
 
-    void setItems(List<DrawerChild> drawerChildren);
+    void setPlaylistItems(List<DrawerChild> drawerChildren);
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerView.java]: This is a good change but probably belongs in a separate commit.

File: app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerPresenter.java [hunk 3]
```diff
@@ -87,7 +89,7 @@ private void closeDrawer() {
         }
     }
 
-    private void loadData() {
+    private void loadData(@NonNull DrawerView drawerView) {
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerPresenter.java]: This is a good change but probably belongs in a separate comit.

File: app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java (hunk 1 above)
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java]: It looks like this was just an oversight - the `Observable` remains subscribed and listening for emissions after we've finished making our playlist menu. We only need the first emission from the playlist relay.

Using the stopEventTrigger/LifecycleCallbacks seems like overkill here. I would think instead, we should just be calling `playlistListObservable.first(Collections.emptyList())`, which ensures the `Observable` only emits once and then is disposed.

File: app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerPresenter.java [hunk 4]
```diff
@@ -111,15 +113,15 @@ public void onOverflowClick(View v, Playlist playlist) {
                                 })
                                 .toList())
                         .observeOn(AndroidSchedulers.mainThread())
+                        .takeUntil(lifecycleProvider.onDestroyed())
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/ui/drawer/DrawerPresenter.java]: This `Observable` is added to a list of `Disposables` via `addDisposable()`, which are disposed of in `unbindView()`. I don't see the benefit of the lifecycle callbacks here - `takeUntil(onDestroyed())` is redundant if we're disposing in `unbindView`.

File: app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+import android.arch.lifecycle.OnLifecycleEvent;
+
+import io.reactivex.Observable;
+import io.reactivex.functions.Function;
+import io.reactivex.subjects.BehaviorSubject;
+
+import static android.arch.lifecycle.Lifecycle.Event.ON_ANY;
+
+/**
+ * Created by Khang NT on 11/29/17.
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java]: Author comments aren't used throughout this app.

File: app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+ * Email: khang.neon.1997@gmail.com
+ */
+
+public class LifecycleProviderHelper implements LifecycleProvider, LifecycleObserver {
+
+    private Lifecycle mHostLifecycle;
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java]: Trying to phase out mNotation throughout the app (use `hostLifecycle` instead)

File: app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    @OnLifecycleEvent(ON_ANY)
+    void onAny(LifecycleOwner source, Lifecycle.Event event) {
+        mSubject.onNext(event);
+        if (event == Lifecycle.Event.ON_DESTROY) {
+            mHostLifecycle.removeObserver(this);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
```
Comment (timusus) [inline app/src/main/java/com/simplecity/amp_library/lifecycle/LifecycleProviderHelper.java]: InheritDoc comments are superfluous.

File: app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java (hunk 1 above)
Comment (Khang-NT) [reply app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java]: Sometime the submenu need to be updated to show new playlist, for instance: in `QueueFragment`, user click submenu -> create playlist, then open menu again, new playlist should be displayed in the menu.

In case `autoUpdate == false`, it goes to `playlistListObservable.take(1)`, equally with 
`playlistListObservable.first(Collections.emptyList())`.

**Comment by Khang-NT** [comment]
Well, `InputMethodManagerLeaks` caused error: `AndroidRuntimeException: requestFeature() must be...`. Seem like this solution will work:
> Use onActivityStarted(Activity) instead of onActivityCreated(Activity, Bundle)

https://github.com/timusus/Shuttle/pull/233/commits/8c7444c93f057e15972e06eee6cf3f8dabc0e707

File: app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java (hunk 1 above)
Comment (timusus) [reply app/src/main/java/com/simplecity/amp_library/utils/PlaylistUtils.java]: The method `makePlaylistMenu()` is there to create the menu, not keep it up-to-date. I think it is an unwanted/accidental side-effect for it to remain subscribed to changes after it has finished creating the menu.

A better solution, IMO, would be to refactor the method, call it `updatePlaylistMenu()` and have it return the `Observable` - then we can keep a reference to the `Observable` and dispose of it in `onPause()`. 

Passing in the Lifecycle Observable still just feels like an over-engineered solution to me. It's harder to understand what's happening. 

I'll do some refactoring of this PR myself and show you what I think is the cleaner solution.

File: libraries/aesthetic/src/main/java/com/afollestad/aesthetic/Aesthetic.java [hunk 8]
```diff
@@ -78,7 +78,7 @@
 
   @SuppressLint("CommitPrefEdits")
   private Aesthetic(Context context) {
-    this.context = context;
+    this.context = context.getApplicationContext();
```
Comment (timusus) [inline libraries/aesthetic/src/main/java/com/afollestad/aesthetic/Aesthetic.java]: This should not be necessary, as long as we're clearing the context reference in `onPause()`

File: libraries/aesthetic/src/main/java/com/afollestad/aesthetic/Aesthetic.java [hunk 9]
```diff
@@ -69,22 +71,33 @@
   private static Aesthetic instance;
 
   private final ArrayMap<String, Integer> lastActivityThemes;
+  private final Context appContext;
 
   private CompositeDisposable subs;
-  private Context context;
+  private WeakReference<Context> weakRefContext;
```
Comment (timusus) [inline libraries/aesthetic/src/main/java/com/afollestad/aesthetic/Aesthetic.java]: I don't really see the purpose of this weak reference. If the context is cleared in onPause(), there's no need for a weak reference.

Comment (Khang-NT) [reply libraries/aesthetic/src/main/java/com/afollestad/aesthetic/Aesthetic.java]: Somehow LeakCanary detect it leak, maybe this condition doesn't work as expected in some specific scenarios:
```java
if (activity.isFinishing()) {
  if (context == activity) {
    context = null;
  }
}
```
Wrap it to weak reference object for safe, since holding activity instance as in static var is not a good design.
